sycl/doc/extensions/LevelZeroBackend/LevelZeroBackend.md

```diff
@@ -293,6 +312,25 @@ should take ownership of the passed native handle. The default behavior is to
 transfer the ownership to the SYCL runtime. See section 4.4 for details. If
 the behavior is "transfer" then the runtime is going to destroy the input
 Level-Zero kernel</td>
+</tr><tr>
+<td>
+
+``` C++
+template <backend Backend, typename T, int Dimensions = 1,
+          typename AllocatorT = buffer_allocator>
+typename std::enable_if<Backend == backend::ext_oneapi_level_zero,
+                        buffer<T, Dimensions, AllocatorT>>::type
+make_buffer(
+    const backend_input_t<backend::ext_oneapi_level_zero,
+                          buffer<T, Dimensions, AllocatorT>> &BackendObject,
+    const context &TargetContext, event AvailableEvent = {})
+```
+</td>
+<td>Constructs a SYCL buffer instance from a Level-Zero memory allocation <code>void *</code>. 

```

Comment:
There are three ways to allocate memory from Level Zero:

* `zeMemAllocShared`
* `zeMemAllocDevice`
* `zeMemAllocHost`

Do we support all three of those cases?  Even if we do support all three, I'd like to see that called out explicitly in the spec.  Maybe something like:

> Construct a SYCL buffer instance from a pointer to a Level Zero memory buffer.  The pointer must be the value returned from a previous call to `zeMemAllocShared()`, `zeMemAllocDevice()`, or `zeMemAllocHost()`.

Note that supporting all three cases would likely require some special handling in the runtime.  For example, a buffer created from `zeMemAllocDevice` would need some logic to copy the memory back to the host in case the application created a `host_accessor` to it.

There should be some description of the "buffer synchronization rules" for a buffer that is created with this interop API.  See [Section 4.7.2.3 "Buffer synchronization rules"](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:buf-sync-rules) of the core SYCL spec.  One option is to say something like this:

> A SYCL buffer that is constructed with this interop API uses the Level Zero memory buffer for its full lifetime, and the contents of the Level Zero memory buffer are unspecified for the lifetime of the SYCL buffer.  If the application modifies the contents of that Level Zero memory buffer during the lifetime of the SYCL buffer, the behavior is undefined.  The initial contents of the SYCL buffer will be the initial contents of the Level Zero memory buffer at the time of the SYCL buffer's construction.
>
> The behavior of the SYCL buffer destructor depends on the `Ownership` flag.  As with other SYCL buffers, this behavior is triggered only when the last reference count to the buffer is dropped, as described in the core SYCL specification section 4.7.2.3, "Buffer synchronization rules".
>
> * If the ownership is `keep` (i.e. the application retains ownership of the Level Zero memory buffer), then the SYCL buffer destructor blocks until all work in queues on the buffer have completed.  The buffer's contents are written to the Level Zero memory buffer by the time the destructor completes.
>
> * If the ownership is `transfer` (i.e. the SYCL runtime has ownership of the Level Zero memory buffer), then the SYCL buffer destructor does not block even if work on the buffer has not completed.  The SYCL runtime frees the Level Zero memory buffer asynchronously when it is no longer in use in queues.

Here are some other things to consider about the behavior:

* It is legal to create an accessor to a buffer that allows access from any device.  For example, a SYCL buffer could be created from Level Zero memory allocated on device *D1*, but it is still legal to create an accessor for that buffer from some other device *D2*.  (The SYCL runtime is responsible for copying memory in this case if necessary.)  You do not list any restriction about this, so I presume it is expected to work even for buffers created with this interop API, correct?

* Are there any issues with creating a sub-buffer from a buffer that is created with this interop API?

Reply:
Due to lack of support in the L0 plugin it is currently not allowed to access buffer from other devices if it is created with zeMemAllocDevice or zeMemAllocShared, I've specified this restriction in the tracker.
It is allowed for buffers created from zeMemAllocHost pointer.

There are no issues with creating sub-buffers. I will update the test with some sub-buffer case.

Reply:
Thanks for the updates.

Since you are talking about tests, it would be good to test this case:

* The buffer is created from memory allocated with `zeMemAllocDevice()`.
* The application creates a `host_accessor` to access the contents of the buffer.

This is an interesting case because the runtime needs to copy the contents of the buffer from the device to the host.

Reply:
Yes, I've added such test to the https://github.com/intel/llvm-test-suite/pull/708